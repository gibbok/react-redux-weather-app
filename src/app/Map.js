import React from 'react'
import NavigationContainer from '../navigation/NavigationContainer'
import MapContainer from '../map/MapContainer'

const Map = () => (
  <div>
    <NavigationContainer />
    <MapContainer />
  </div>
)

export default Map
