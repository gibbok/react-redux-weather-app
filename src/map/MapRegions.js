import React from 'react'

const MapRegions = ({ regions, onRegionsChange }) => {
  const options = regions.map((region) =>
    <option
      value={region.id}
      key={region.id}
    >{region.title}
    </option>
  )
  return (
    <select onChange={(e) => onRegionsChange(e.target.value)}>
      {options}
    </select>
  )
}

export default MapRegions
