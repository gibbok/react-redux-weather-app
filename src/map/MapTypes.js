import React from 'react'

const MapTypes = ({ types, onTypesClick }) => {
  return (
    <div >
      {types.map((type) =>
        <div onClick={(e) => onTypesClick(type.id)}
          key={type.id}
          >{type.title}</div>)}
    </div>
  )
}

export default MapTypes
