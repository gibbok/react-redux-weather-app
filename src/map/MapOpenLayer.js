import React from 'react'
import * as ol from 'openlayers'
import * as api from '../app/api'

let mapOpenLayer = {
  map: undefined,
  init (props) {
    let geo = props.geo

      // render marker vector
    let markerFeature = new ol.Feature({
      geometry: new ol.geom.Point(ol.proj.transform([-72.0704, 46.678], 'EPSG:4326', 'EPSG:3857')) // TODO // take lat long from openweather api which should be sotred in the state
    })

    let markerSource = new ol.source.Vector({
      features: [markerFeature]
    })

    let markerStyle = new ol.style.Style({
      image: new ol.style.Icon(({
        anchor: [0, 0],
        anchorXUnits: 'fraction',
        anchorYUnits: 'pixels',
        opacity: 0.75,
        src: 'assets/pin.svg'
      }))
    })

    let markerLayer = new ol.layer.Vector({
      source: markerSource,
      style: markerStyle
    })

      // render OpenStreetMap tile server
    let tileLayer = new ol.layer.Tile({
      source: new ol.source.OSM()
    }, new ol.layer.Vector({
      source: new ol.source.Vector({ features: [], projection: 'EPSG:4326' })
    }))
      // create map
    this.map = new ol.Map({
      target: 'map',
      layers: [
        tileLayer,
        markerLayer,
        new ol.layer.Tile({
          source: new ol.source.XYZ({
            url: api.mapTemperature()
          })
        })
      ],
      view: new ol.View({
        center: ol.proj.transform(geo, 'EPSG:4326', 'EPSG:3857'),
        zoom: 4
      })
    })
  },
  changeLayer (name) {
    let url
    switch (name) {
      case 'default':
      case 'temperature':
        url = api.mapTemperature() // ok
        break
      case 'precipitation':
        url = api.mapPrecipitation()
        break
      case 'pressure':
        url = api.mapPressure()
        break
      case 'wind':
        url = api.mapWind()
        break
      case 'cloud':
        url = api.mapCloud()
        break
    }
    let tile = new ol.source.XYZ({
      url: url
    })
    let layer = this.map.getLayers().getArray()[2]
    layer.setSource(tile)
  }
}

const MapOpenLayer = React.createClass({
  getInitialState () {
    return {
      activeRegion: 'currentLocation',
      activeType: 'temperature'
    }
  },
  componentDidMount () {
    mapOpenLayer.init(this.props)
  },
  componentDidUpdate (prevProps) {
    let activeRegion = Object.assign({}, this.props.regions.find(region => region.isActive))
    let activeType = Object.assign({}, this.props.types.find(mapType => mapType.isActive))

    if (activeRegion.id !== this.state.activeRegion) {
      this.animateMapToRegion()
      this.state.activeRegion = activeRegion.id
    }

    if (activeType.id !== this.state.activeType) {
      mapOpenLayer.changeLayer(activeType.id)
      this.state.activeType = activeType.id
    }
  },
  animateMapToRegion () {
    let nextActiveRegion = this.props.regions.find(region => region.isActive)
    let newCenterMap
    switch (nextActiveRegion.id) {
      case 'currentLocation':
        newCenterMap = this.props.geo
        break
      case 'northAmerica':
        newCenterMap = [-100.7838, 46.8083]
        break
      case 'southAmerica':
        newCenterMap = [-57.6534, -19.0092]
        break
      case 'europe':
        newCenterMap = [8.6833, 50.1167]
        break
      case 'asia':
        newCenterMap = [91.1, 29.65]
        break
      case 'pacificIslands':
        newCenterMap = [156.3159, -7.5559]
        break
      case 'africa':
        newCenterMap = [18.5333, 4.3833]
        break
    }

    let view = mapOpenLayer.map.getView()
    view.animate({
      center: ol.proj.fromLonLat(newCenterMap),
      duration: 1000,
      zoom: 4
    })
  },
  render () {
    return (
      <div id='map' className='map' />
    )
  }
})

export default MapOpenLayer
