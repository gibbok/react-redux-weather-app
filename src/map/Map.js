import React from 'react'
import * as ol from 'openlayers'
import RegionsSelect from './RegionsSelect'
import TypesSelect from './TypesSelect'
import MapOpenLayer from './MapOpenLayer'
import styles from '../../node_modules/openlayers/css/ol.css' // eslint-disable-line no-unused-vars

const Map = React.createClass({
  // componentDidUpdate (prevProps) {
  //   this.animateMapToRegion(prevProps)
  // },
  // animateMapToRegion (prevProps) {
  //   let nextActiveRegion = prevProps.map.regions.find(region => region.isActive)
  //   let newCenterMap
  //   switch (nextActiveRegion.id) {
  //     case 'currentLocation':
  //       newCenterMap = this.props.geo
  //       break
  //     case 'northAmerica':
  //       newCenterMap = [-100.7838, 46.8083]
  //       break
  //     case 'southAmerica':
  //       newCenterMap = [-57.6534, -19.0092]
  //       break
  //     case 'europe':
  //       newCenterMap = [8.6833, 50.1167]
  //       break
  //     case 'asia':
  //       newCenterMap = [91.1, 29.65]
  //       break
  //     case 'pacificIslands':
  //       newCenterMap = [156.3159, -7.5559]
  //       break
  //     case 'africa':
  //       newCenterMap = [18.5333, 4.3833]
  //       break
  //   }

  //   let view = mapOpenLayer.map.getView()
  //   view.animate({
  //     center: ol.proj.fromLonLat(newCenterMap),
  //     duration: 1000,
  //     zoom: 4
  //   })
  // },
  render () {
    return (
      <div>
        <RegionsSelect
          regions={this.props.regions}
          onRegionsChange={this.props.onRegionsChange}
         />
        <TypesSelect
          types={this.props.types}
          onTypesChange={this.props.onTypesChange}
          />
        <MapOpenLayer
          geo={this.props.geo}
          regions={this.props.regions}
          types={this.props.types}
          />
      </div>
    )
  }
})

export default Map
