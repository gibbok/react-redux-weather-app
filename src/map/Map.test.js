import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Map from './Map'
import MapRegions from './MapRegions'
import MapTypes from './MapTypes'
import MapOpenLayer from './MapOpenLayer'

describe('<Map />', () => {
  it('should render', () => {
    const data = {
      regions: [
        {
          id: 'currentLocation',
          title: 'Current Location',
          isActive: true
        },
        {
          id: 'northAmerica',
          title: 'North America',
          isActive: false
        }
      ],
      types: [
        {
          id: 'temperature',
          title: 'Temperature',
          isActive: true
        },
        {
          id: 'precipitation',
          title: 'Precipitation',
          isActive: false
        }
      ],
      geo: [-74.006, 40.7143]
    }

    const wrapper = shallow(
      <Map
        regions={data.regions}
        types={data.types}
        onTypesChange={() => {}}
        onRegionsChange={() => {}}
        geo={data.geo}
        />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should contain MapTypes, MapRegions and MapOpenLayer components', () => {
    const data = {
      regions: [
        {
          id: 'currentLocation',
          title: 'Current Location',
          isActive: true
        },
        {
          id: 'northAmerica',
          title: 'North America',
          isActive: false
        }
      ],
      types: [
        {
          id: 'temperature',
          title: 'Temperature',
          isActive: true
        },
        {
          id: 'precipitation',
          title: 'Precipitation',
          isActive: false
        }
      ]
    }

    const wrapper = shallow(
      <Map
        regions={data.regions}
        types={data.types}
        onTypesChange={() => {}}
        onRegionsChange={() => {}}
        geo={[-74.006, 40.7143]}
          />
      )
    expect(wrapper.find(MapTypes).length).toBe(1)
    expect(wrapper.find(MapRegions).length).toBe(1)
    expect(wrapper.find(MapOpenLayer).length).toBe(1)
  })
})
