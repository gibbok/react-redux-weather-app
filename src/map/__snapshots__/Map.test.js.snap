// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> should render 1`] = `
<div>
  <MapTypes
    onTypesClick={[Function]}
    types={
      Array [
        Object {
          "id": "temperature",
          "isActive": true,
          "title": "Temperature",
        },
        Object {
          "id": "precipitation",
          "isActive": false,
          "title": "Precipitation",
        },
      ]
    }
  />
  <MapRegions
    onRegionsChange={[Function]}
    regions={
      Array [
        Object {
          "id": "currentLocation",
          "isActive": true,
          "title": "Current Location",
        },
        Object {
          "id": "northAmerica",
          "isActive": false,
          "title": "North America",
        },
      ]
    }
  />
  <MapOpenLayer
    geo={
      Array [
        -74.006,
        40.7143,
      ]
    }
    regions={
      Array [
        Object {
          "id": "currentLocation",
          "isActive": true,
          "title": "Current Location",
        },
        Object {
          "id": "northAmerica",
          "isActive": false,
          "title": "North America",
        },
      ]
    }
    types={
      Array [
        Object {
          "id": "temperature",
          "isActive": true,
          "title": "Temperature",
        },
        Object {
          "id": "precipitation",
          "isActive": false,
          "title": "Precipitation",
        },
      ]
    }
  />
</div>
`;
